include/devicetree.h

```diff
@@ -1714,6 +1714,105 @@
 #define DT_FOREACH_PROP_ELEM_VARGS(node_id, prop, fn, ...)		\
 	DT_CAT4(node_id, _P_, prop, _FOREACH_PROP_ELEM_VARGS)(fn, __VA_ARGS__)
 
+/**
+ * @brief Call "fn" on all nodes with compatible DT_DRV_COMPAT
+ *        and status "okay"
+ *
+ * This macro expands to:
+ *
+ *     fn(node_id_1) fn(node_id_2) ... fn(node_id_n)
+ *
+ * where each "node_id_<i>" is a node identifier for some node with
+ * compatible "compat" and status "okay". Whitespace is added between
+ * expansions as shown above.
+ *
+ * Example devicetree fragment:
+ *
+ *     / {
+ *             a {
+ *                     compatible = "foo";
+ *                     status = "okay";
+ *             };
+ *             b {
+ *                     compatible = "foo";
+ *                     status = "disabled";
+ *             };
+ *             c {
+ *                     compatible = "foo";
+ *             };
+ *     };
+ *
+ * Example usage:
+ *
+ *     DT_FOREACH_STATUS_OKAY(foo, DT_NODE_PATH)
+ *
+ * This expands to one of the following:
+ *
+ *     "/a" "/c"
+ *     "/c" "/a"
+ *
+ * "One of the following" is because no guarantees are made about the
+ * order that node identifiers are passed to "fn" in the expansion.
+ *
+ * (The "/c" string literal is present because a missing status
+ * property is always treated as if the status were set to "okay".)
+ *
+ * Note also that "fn" is responsible for adding commas, semicolons,
+ * or other terminators as needed.
+ *
+ * @param compat lowercase-and-underscores devicetree compatible
+ * @param fn Macro to call for each enabled node. Must accept a
+ *           node_id as its only parameter.
+ */
+#define DT_FOREACH_STATUS_OKAY(compat, fn)				\
```

Comment:
`DT_INST_FOREACH_STATUS_OKAY` could now use this, right?

Reply:
> `DT_INST_FOREACH_STATUS_OKAY` could now use this, right?

I thought so too originally, but the answer is surprisingly no.

The reason is that "fn" as passed to DT_INST_FOREACH_STATUS_OKAY expects an instance number. As it's written here, though, "fn" as passed to DT_FOREACH_STATUS_OKAY expects a node_id.

So for DT_INST_FOREACH_STATUS_OKAY to use DT_FOREACH_STATUS_OKAY, we would need some way to convert a node_id to an instance number of a compatible. But we don't have such a thing. We could add something like that -- like a `DT_NODE_INSTANCE_NUMBER(node_id, compat)` macro that returns the node's instance number when considered as a `compat`. But that would add another intermediate macro call in the expansion of DT_INST_FOREACH_STATUS_OKAY, for every single call to `fn`.

I think that would make debugging more complicated. I designed the way the for-each helper macros in devicetree_unfixed.h work to minimize the number of intermediate macro calls. This makes the GCC error messages shorter and easier to understand when there are user errors in `fn` or `compat` etc. That's why we don't use FOR_EACH or any util_macro.h calls also -- the error messages they blow up into if you misuse LOG_ERR or whatever are really hard to deal with.

devicetree.h is already hard enough to use, so I think prioritizing ease of use at the cost of a few extra lines of macros is a good tradeoff.

Reply:
My above comment  is a longer version of what I am trying to say in this comment in one of the patches:

```
        # Helpers for INST versions of for-each macros, which take
        # instance numbers. We emit separate helpers for these because
        # avoiding an intermediate node_id --> instance number
        # conversion in the preprocessor helps to keep the macro
        # expansions simpler. That hopefully eases debugging.
```